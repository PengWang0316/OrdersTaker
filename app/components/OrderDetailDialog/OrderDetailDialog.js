import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Dialog, DialogContent, Button, IconButton } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { Close } from '@material-ui/icons/';

import dialogStyles from '../../styles/DialogSytels';
import OrderBrowserCardHeader from '../OrderBrowser/OrderBrowserCardHeader';
import CartPageContainer from '../containers/CartPageContainer';
import OrderList from '../OrderList';
import OrderSummaryCategories from '../OrderSummaryCategories';
import OrderSummaryPrice from '../OrderSummaryPrice';
import ItemDetailDialog from '../ItemDetailDialog';
import ShowDetailDialogContext from '../../contexts/ShowDetailDialogContext'; // Import the context to pass the function.


const styles = ({
  ...dialogStyles, // Extending some basic dialog styles from DialogSytels.
  dialogPaper: { ...dialogStyles.dialogPager, minWidth: 300, maxWidth: 1100 },
  headerDiv: { // These styles are using to cancel the CardHeader's default padding styles.
    position: 'relative',
    top: -20,
    left: -30
  },
  contentDiv: {
    display: 'grid',
    gridTemplateColums: '1fr 300px',
    gridTemplateRows: '1fr',
    gridTemplateAreas: '"contentPanel summaryPanel"',
    gridGap: '2em'
  },
  contentPanel: {
    gridArea: 'contentPanel'
  },
  summaryPanel: {
    gridArea: 'summaryPanel',
    minWidth: 240
  },
  summaryContent: {
    position: 'sticky',
    top: 20
  },
  priceDiv: {
    display: 'flex',
    contentJustify: 'flex-end'
  },
  '@media (max-width: 734px)': {
    contentDiv: {
      gridTemplateColums: '1fr',
      gridTemplateRows: '100% 1fr',
      gridTemplateAreas: `"summaryPanel"
                          "contentPanel"`
    }
  }
});

/**
 * The component that is using to show the order's detail.
 */
export class OrderDetailDialog extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    onClose: PropTypes.func.isRequired,
    open: PropTypes.bool.isRequired,
    order: PropTypes.object,
    menuItems: PropTypes.object
  };

  static defaultProps = {
    order: null,
    menuItems: null
  };

  state = {
    isItemDetailDialogOpen: false,
    currentOrder: null
  }

  /**
   * Setting the isItemDetailDialogOpen state to the opposite value.
   * @return {null} No return.
   */
  handleDialogToggle = () => this.setState(({ isItemDetailDialogOpen }) => ({ isItemDetailDialogOpen: !isItemDetailDialogOpen }));

  /**
   * Setting up a new currentOrder state and call the handleDialogToggle method.
   * @param {string} orderId is the id of new order will be showed in the detail dialog.
   * @return {null} No return.
   */
  showDetailDialog = orderId => this.setState({ currentOrder: this.props.menuItems[orderId] }, () => this.handleDialogToggle());

  /**
   * The render method.
   * @return {jsx} Returning the jsx for the component.
   */
  render() {
    const { onClose, open, order, menuItems, classes } = this.props;
    const { currentOrder, isItemDetailDialogOpen } = this.state;
    let newOrder;
    if (this.lastOrder === order && this.lastNewOrder) newOrder = this.lastNewOrder; // If the order did not change and we have the computed new order, just return instead of recomputing.
    else if (order) {
      newOrder = CartPageContainer.parseOrders(order.items, menuItems); // Calling the parserOrders method from the CartPageContainer component.
      this.lastNewOrder = newOrder; // Storaging it for caching.
      this.lastOrder = order; // Storaging the current order to a veriable.
    }
    return (
      <Fragment>
        <Dialog onClose={onClose} open={open} classes={{ paper: classes.dialogPaper }}>
          <DialogContent className={classes.dialogContent}>
            {newOrder && (
              <Fragment>
                <IconButton onClick={onClose} className={classes.dialogCloseButton}><Close className={classes.dialogCloseIcon} /></IconButton>
                <div className={classes.headerDiv}><OrderBrowserCardHeader order={order} /></div>
                <div className={classes.contentDiv}>
                  <div className={classes.contentPanel}>
                    <ShowDetailDialogContext.Provider value={this.showDetailDialog}>
                      <OrderList orders={newOrder} orderItems={order.items} isHistoryOrder />
                    </ShowDetailDialogContext.Provider>
                  </div>
                  <div className={classes.summaryPanel}>
                    <div className={classes.summaryContent}>
                      <OrderSummaryCategories orders={newOrder} />
                      <div classes={classes.priceDiv}>
                        <OrderSummaryPrice orders={newOrder} />
                      </div>
                    </div>
                  </div>
                </div>
                <div className={classes.dialogButtonDiv}>
                  <Button onClick={onClose} variant="contained" size="small" aria-label="Close the detail page" color="primary">Close</Button>
                </div>
              </Fragment>
            )}
          </DialogContent>
        </Dialog>
        {newOrder && <ItemDetailDialog onClose={this.handleDialogToggle} open={isItemDetailDialogOpen} item={currentOrder} />}
      </Fragment>
    );
  }
}
export default withStyles(styles)(OrderDetailDialog);
