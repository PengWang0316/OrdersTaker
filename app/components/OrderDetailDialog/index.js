import OrderDetailDialog from './OrderDetailDialog';

export default OrderDetailDialog;
