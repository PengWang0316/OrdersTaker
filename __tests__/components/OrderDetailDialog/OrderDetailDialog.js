import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { OrderDetailDialog } from '../../../app/components/OrderDetailDialog/OrderDetailDialog';

jest.mock('@material-ui/core/Dialog', () => 'Dialog');
jest.mock('@material-ui/core/DialogContent', () => 'DialogContent');
jest.mock('@material-ui/core/Button', () => 'Button');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/icons/Close', () => 'Close');
jest.mock('../../../app/components/OrderBrowser/OrderBrowserCardHeader', () => 'OrderBrowserCardHeader');
jest.mock('../../../app/components/containers/CartPageContainer', () => ({ parseOrders: jest.fn().mockReturnValue('returnOrders') }));
jest.mock('../../../app/components/OrderList', () => 'OrderList');
jest.mock('../../../app/components/OrderSummaryCategories', () => 'OrderSummaryCategories');
jest.mock('../../../app/components/OrderSummaryPrice', () => 'OrderSummaryPrice');
jest.mock('../../../app/components/ItemDetailDialog', () => 'ItemDetailDialog');

describe('OrderDetailDialog', () => {
  const defaultProps = {
    classes: {
      headerDiv: 'hearderDiv',
      contentDiv: 'contentDiv',
      contentPanel: 'contentPanel',
      summaryPanel: 'summayPanel',
      summayContent: 'summayContent',
      priceDiv: 'priceDiv'
    },
    onClose: jest.fn(),
    open: true
  };
  const getShallowComponent = (props = defaultProps) => shallow(<OrderDetailDialog {...props} />);

  test('Initial states and part of cache functionality', () => {
    const CartPageContainer = require('../../../app/components/containers/CartPageContainer');
    const component = getShallowComponent();
    const { isItemDetailDialogOpen, currentOrder } = component.state();
    expect(isItemDetailDialogOpen).toBe(false);
    expect(currentOrder).toBeNull();
    expect(component.instance().lastNewOrder).toBeUndefined();
    expect(CartPageContainer.parseOrders).not.toHaveBeenCalled();
  });

  test('The cache functionality', () => {
    const CartPageContainer = require('../../../app/components/containers/CartPageContainer');
    const component = getShallowComponent({ ...defaultProps, order: { items: 'orderItems' }, menuItems: { id: 'menuItems' } });
    expect(CartPageContainer.parseOrders).toHaveBeenCalledTimes(1);
    expect(CartPageContainer.parseOrders).toHaveBeenLastCalledWith('orderItems', { id: 'menuItems' });
    expect(component.instance().lastNewOrder).toBe('returnOrders');
    expect(component.instance().lastOrder).toEqual({ items: 'orderItems' });

    // Calling the render method again to test the cache functionality.
    component.instance().render();
    expect(CartPageContainer.parseOrders).toHaveBeenCalledTimes(1);
  });

  test('handleDialogToggle', () => {
    const component = getShallowComponent();
    expect(component.state('isItemDetailDialogOpen')).toBe(false);
    component.instance().handleDialogToggle();
    expect(component.state('isItemDetailDialogOpen')).toBe(true);
  });

  test('showDetailDialog', () => {
    const mockHandleDialogToggleFn = jest.fn();
    const component = getShallowComponent({ ...defaultProps, menuItems: { id: 'menuItems' } });
    component.instance().handleDialogToggle = mockHandleDialogToggleFn;
    expect(component.state('currentOrder')).toBeNull();
    component.instance().showDetailDialog('id');
    expect(component.state('currentOrder')).toBe('menuItems');
    expect(mockHandleDialogToggleFn).toHaveBeenCalledTimes(1);
  });

  test('Snapshot without newOrder', () => expect(renderer.create(<OrderDetailDialog {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot with newOrder', () => expect(renderer.create(<OrderDetailDialog {...{ ...defaultProps, order: { items: 'orderItems' }, menuItems: { id: 'menuItems' } }} />).toJSON()).toMatchSnapshot());
});
