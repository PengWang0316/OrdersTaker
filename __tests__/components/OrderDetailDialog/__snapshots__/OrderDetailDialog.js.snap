// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderDetailDialog Snapshot with newOrder 1`] = `
Array [
  <Dialog
    classes={
      Object {
        "paper": undefined,
      }
    }
    onClose={[MockFunction]}
    open={true}
  >
    <DialogContent>
      <IconButton
        onClick={[MockFunction]}
      >
        <Close />
      </IconButton>
      <div
        className="hearderDiv"
      >
        <OrderBrowserCardHeader
          order={
            Object {
              "items": "orderItems",
            }
          }
        />
      </div>
      <div
        className="contentDiv"
      >
        <div
          className="contentPanel"
        >
          <OrderList
            isHistoryOrder={true}
            orderItems="orderItems"
            orders="returnOrders"
          />
        </div>
        <div
          className="summayPanel"
        >
          <div>
            <OrderSummaryCategories
              orders="returnOrders"
            />
            <div
              classes="priceDiv"
            >
              <OrderSummaryPrice
                orders="returnOrders"
              />
            </div>
          </div>
        </div>
      </div>
      <div>
        <Button
          aria-label="Close the detail page"
          color="primary"
          onClick={[MockFunction]}
          size="small"
          variant="contained"
        >
          Close
        </Button>
      </div>
    </DialogContent>
  </Dialog>,
  <ItemDetailDialog
    item={null}
    onClose={[Function]}
    open={false}
  />,
]
`;

exports[`OrderDetailDialog Snapshot without newOrder 1`] = `
<Dialog
  classes={
    Object {
      "paper": undefined,
    }
  }
  onClose={[MockFunction]}
  open={true}
>
  <DialogContent />
</Dialog>
`;
